embassy/src/util/mpsc.rs

```diff
@@ -0,0 +1,917 @@
+
+impl<'ch, T> Sender<'ch, T> {
+    /// Sends a value, waiting until there is capacity.
+    ///
+    /// A successful send occurs when it is determined that the other end of the
+    /// channel has not hung up already. An unsuccessful send would be one where
+    /// the corresponding receiver has already been closed. Note that a return
+    /// value of `Err` means that the data will never be received, but a return
+    /// value of `Ok` does not mean that the data will be received. It is
+    /// possible for the corresponding receiver to hang up immediately after
+    /// this function returns `Ok`.
+    ///
+    /// # Errors
+    ///
+    /// If the receive half of the channel is closed, either due to [`close`]
+    /// being called or the [`Receiver`] handle dropping, the function returns
+    /// an error. The error includes the value passed to `send`.
+    ///
+    /// [`close`]: Receiver::close
+    /// [`Receiver`]: Receiver
+    pub async fn send(&self, message: T) -> Result<(), SendError<T>> {
+        SendFuture {
+            sender: self.clone(),
+            message: UnsafeCell::new(message),
+        }
+        .await
+    }
+
+    /// Attempts to immediately send a message on this `Sender`
+    ///
+    /// This method differs from [`send`] by returning immediately if the channel's
+    /// buffer is full or no receiver is waiting to acquire some data. Compared
+    /// with [`send`], this function has two failure cases instead of one (one for
+    /// disconnection, one for a full buffer).
+    ///
+    /// # Errors
+    ///
+    /// If the channel capacity has been reached, i.e., the channel has `n`
+    /// buffered values where `n` is the argument passed to [`channel`], then an
+    /// error is returned.
+    ///
+    /// If the receive half of the channel is closed, either due to [`close`]
+    /// being called or the [`Receiver`] handle dropping, the function returns
+    /// an error. The error includes the value passed to `send`.
+    ///
+    /// [`send`]: Sender::send
+    /// [`channel`]: channel
+    /// [`close`]: Receiver::close
+    pub fn try_send(&self, message: T) -> Result<(), TrySendError<T>> {
+        unsafe { &mut *self.channel.get() }.try_send(message)
+    }
+
+    /// Completes when the receiver has dropped.
+    ///
+    /// This allows the producers to get notified when interest in the produced
+    /// values is canceled and immediately stop doing work.
+    pub async fn closed(&self) {
+        CloseFuture {
+            sender: self.clone(),
+        }
+        .await
+    }
+
+    /// Checks if the channel has been closed. This happens when the
+    /// [`Receiver`] is dropped, or when the [`Receiver::close`] method is
+    /// called.
+    ///
+    /// [`Receiver`]: crate::sync::mpsc::Receiver
+    /// [`Receiver::close`]: crate::sync::mpsc::Receiver::close
+    pub fn is_closed(&self) -> bool {
+        unsafe { &mut *self.channel.get() }.is_closed()
+    }
+}
+
+struct SendFuture<'ch, T> {
```

Comment:
Wdyt about making this type public, along with a pub fn new() that accepts a Sender instance and a message to be sent? This should allow for the use case where the future type needs to be known in advance.

Reply:
Why have a public `new` though?

Reply:
The type should be opaque to reduce the surface area, right?

Reply:
@huntc To avoid needing to use the Sender.send() function for cases that need to know the future type. For instance, so one can use the channel in code like this:

```rust
struct MyStruct {
   future: SendFuture<'static, MyMessage>
}

fn foo(sender: Sender<'static, MyMessage>) -> MyStruct {
     MyStruct {
        future: SendFuture::new(sender.clone(), MyMessage)
    }
}
```

I'm not sure how I can store the future returned from Sender.send() for later use otherwise without needing alloc.